spec/mongoid/clients/transactions_spec.rb

```diff
@@ -0,0 +1,373 @@
+  let(:subscriber) do
+    Mongoid::Clients.with_name(:other).send(:monitoring).subscribers['Command'].find do |s|
+      s.is_a?(EventSubscriber)
+    end
+  end
+
+  let(:insert_events) do
+    subscriber.started_events.select { |event| event.command_name == 'insert' }
+  end
+
+  let(:insert_events_txn_numbers) do
+    insert_events.map { |i| i.instance_variable_get(:@integer) }
+  end
+
+  let(:update_events) do
+    subscriber.started_events.select { |event| event.command_name == 'update' }
+  end
+
+  let(:update_events_txn_numbers) do
+    update_events.map { |i| i.instance_variable_get(:@integer) }
+  end
+
+  let(:other_events) do
+    subscriber.started_events.reject { |event| ['insert', 'update'].include?(event.command_name) }
+  end
+
+  context 'when a transaction is used on a model class' do
+
+    context 'when transactions are supported', if: testing_transactions? do
+
+      around do |example|
+        Mongoid::Clients.with_name(:other).database.collections.each(&:drop)
+        Mongoid::Clients.with_name(:other).command(create: :people)
+        Mongoid::Clients.with_name(:other).command(create: :posts)
+        subscriber.clear_events!
+        Person.with(client: :other) do
+          example.run
+        end
+        Mongoid::Clients.with_name(:other).database.collections.each(&:drop)
+      end
+
+      context 'when another thread is started' do
+
+        let!(:last_use_diff) do
+          Person.with_session do
+            Person.with_transaction do
+              Person.create
+              Person.create
+              Thread.new { Person.create }.value
+            end
+          end
+        end
+
+        it 'does not use the transaction for that thread' do
+          expect(Person.count).to be(2)
+          expect(Person.with(client: :default) { Person.count }).to be(1)
+          expect(insert_events.count { |e| e.command['startTransaction'] }).to be(1)
+          expect(other_events.count { |e| e.command_name == 'commitTransaction' }).to be(1)
+        end
+      end
+
+      context 'when the operations in the transactions block are all on the class' do
+
+        before do
+          Person.with_session do
+            Person.with_transaction do
+              Person.create
+              Person.create
+            end
+          end
+        end
+
+        it 'uses a single transaction number for all operations on the class' do
+          expect(Person.count).to be(2)
+          expect(insert_events_txn_numbers.size).to eq(2)
+          expect(insert_events_txn_numbers.uniq.size).to eq(1)
+        end
+      end
+
+      context 'when the operations in the transactions block are also on another class' do
+
+        context 'when the other class uses the same client' do
+
+          before do
+            Post.with(client: :other) do
+              Person.with_session do
+                Person.with_transaction do
+                  Person.create
+                  Person.create
+                  Post.create
```

Comment:
There needs to be another test added which:
1. Creates these 3 objects
2. Queries server for these 3 objects, verifies that they are returned.
3. Raises an exception
4. Outside of the transaction block, queries the server again for these 3 objects, verifies that none of them is returned.